### PR TITLE
Preserve xcuserdata when re-generating projects

### LINKED ISSUE #1003: Preserve Xcode user data when re-generating projects & workspaces
**Describe the bug**

Custom Xcode UI settings are lost when re-generating projects or workspaces. For example, breakpoints, last opened file, last selected scheme etc... This sadly impacts developer workflow when switching over to using project generation.

**To Reproduce**

1. Generate any project within `fixtures`
1. Open the generated project
1. Add breakpoints, select a source file, navigate through various source files
1. Close Xcode
1. Re-open the project
1. Notice the last viewed file is still open, the recently opened files are also still there along with the breakpoints
1. Close Xcode
1. Re-generate the project
1. Open the generated project
1. Notice the last opened file, recently opened files and break points are lost

**Expected behavior**
The last opened file, recently opened files and break points are preserved.

**Desktop (please complete the following information):**

- Version 1.2.0

**Additional context**

The UI state of Xcode (e.g. breakpoints state, last opened files, window sizes etc...) is stored within a directory called `xcuserdata` which is typically git ignored. This lives within the `.xcodeproj` and `.xcworkspace` directories. 

```
App.xcworkspace/
├── contents.xcworkspacedata
├── xcshareddata
│   └── IDEWorkspaceChecks.plist
└── xcuserdata
    └── kwridan.xcuserdatad
        └── UserInterfaceState.xcuserstate
```

To solve this we can replace a selection of files/directories rather than the entire `.xcodeproj` and `.xcworkspace` directories:

- For projects: `.xcodeproj/project.pbxproj` and the `xcshareddata/xcschemes` directory 
- For workspaces: `.xcworkspace/contents.xcworkspacedata` and the `.xcworkspace/xcshareddata/xcschemes` directory

### PR DESCRIPTION
**Pull request by kwridan**
Resolves #1003 

### Short description 📝
- The project generation process was replacing the entire `.xcodeproj` and `.xcworkspace` directories
- This lead to wiping all previous contents within those directories
- This included a nested `xcuserdata` directory which is usually gitignored and holds the Xcode UI state

### Solution 📦

- To resolve this, only a the individual files are replaced rather than the entire directory

### Implementation 👩‍💻👨‍💻

- [x] Update project generator
- [x] Update workspace generator
- [x] Update changelog

### Test Plan 🛠:

- Run `tuist generate` within any fixture
- Open the generated project
- Navigate around, set breakpoints
- Re-order one of the source files (i.e. to cause a re-generation)
- Close Xcode
- Run `tuist generate` again
- Open the generated project
- Verify the last viewed file is opened automatically, breakpoints, the recently viewed files list are all still there

### CODE REVIEW

**Comment by kwridan** [comment]
Actually my approach here has a flaw ..

For any schemes removed, they won't get deleted when re-generating. Will need to think of another approach, perhaps more explicit listing of what should be replaced instead of `/**`

e.g. 
```swift
let pathsToReplace = [
   "project.pbxproj",
   "contents.xcworkspace",
   "xcshareddata",
   "xcuserdata/xcschemes/**"
]
```

File: Sources/TuistGenerator/Generator/ProjectGenerator.swift [hunk 1]
```diff
@@ -263,25 +263,51 @@ final class ProjectGenerator: ProjectGenerating {
                        pbxproj: PBXProj,
                        project: Project,
                        graph: Graphing) throws -> GeneratedProject {
-        var generatedProject: GeneratedProject!
-
-        try FileHandler.shared.inTemporaryDirectory { temporaryPath in
-
+        let fileHandler = FileHandler.shared
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/ProjectGenerator.swift]: I don't think we need to define this here. We can access `FileHandler.shared` directly.

File: Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift [hunk 2]
```diff
@@ -48,6 +48,70 @@ final class ProjectGeneratorTests: TuistUnitTestCase {
...
+        let project = Project.test(path: temporaryPath, name: "Foo", targets: [target])
+
+        let graph = Graph.create(project: project,
+                                 dependencies: [
+                                     (target: target, dependencies: []),
+                                 ])
+
+        // When
+        try (0 ..< 2).forEach { _ in
+            _ = try subject.generate(project: project, graph: graph)
```
Comment (pepibumur) [inline Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift]: What about moving theses tests to `ProjectGeneratorIntegrationTests`. Based on the implementation it seems that we are not stubbing the calls to the subject's dependencies.

File: Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift [hunk 3]
```diff
@@ -48,6 +48,70 @@ final class ProjectGeneratorTests: TuistUnitTestCase {
...
+        }
+
+        // Then
+        XCTAssertTrue(paths.allSatisfy { FileHandler.shared.exists($0) })
+    }
+
+    func test_generate_replacesSchemes() throws {
```
Comment (pepibumur) [inline Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift]: Same comment here in regards to where the test should be defined.

File: Tests/TuistGeneratorTests/Generator/WorkspaceGeneratorTests.swift [hunk 4]
```diff
@@ -80,6 +80,29 @@ final class WorkspaceGeneratorTests: TuistUnitTestCase {
         )
     }
 
+    func test_generate_doesNotWipeUserData() throws {
```
Comment (pepibumur) [inline Tests/TuistGeneratorTests/Generator/WorkspaceGeneratorTests.swift]: Same comment here in regards to where the test should be defined.

File: Sources/TuistGenerator/Generator/ProjectGenerator.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistGenerator/Generator/ProjectGenerator.swift]: The intention was to reduce some of the line lengths below where it was used

File: Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift (hunk 2 above)
Comment (kwridan) [reply Tests/TuistGeneratorTests/Generator/ProjectGeneratorTests.swift]: makes sense, I believe the entire file can be moved there as the majority of tests here appear to be using the default implementations.

**Comment by pepibumur** [comment]
Your changes look good to me @kwridan

**Pull request merged**